game/phase.go

```diff
@@ -1211,6 +1222,82 @@ func (p *PhaseResolver) Act() error {
 	return nil
 }
 
+func (p *PhaseResolver) provideGrace(orderMap map[godip.Nation]map[godip.Province][]string) (bool, error) {
+	if p.Game.GracePeriodMinutes == 0 || p.Game.GracePeriodsPerPlayer == 0 {
+		return false, nil
+	}
+
+	provideGrace := false
+	graceNations := []string{}
+	for memberIdx, member := range p.Game.Members {
+		_, hadOrders := orderMap[member.Nation]
+		wasReady := false
+		for _, phaseState := range p.PhaseStates {
+			if phaseState.Nation == member.Nation {
+				wasReady = phaseState.ReadyToResolve
+			}
+		}
+		if !wasReady && !hadOrders && member.GracePeriodsUsed < p.Game.GracePeriodsPerPlayer {
+			p.Game.Members[memberIdx].GracePeriodsUsed++
+			provideGrace = true
+			graceNations = append(graceNations, string(member.Nation))
+		}
+	}
+
+	if provideGrace {
+		// Postpone the phase.
+		now := time.Now()
+		p.Phase.DeadlineAt = now.Add(time.Minute * p.Game.GracePeriodMinutes)
+
+		// Save everything.
+		phaseID, err := p.Phase.ID(p.Context)
+		if err != nil {
+			log.Errorf(p.Context, "p.Phase.ID(...): %v; wtf?", err)
+			return false, err
+		}
+		toSave := []interface{}{
+			p.Game, p.Phase,
+		}
+		keys := []*datastore.Key{
+			p.Game.ID, phaseID,
+		}
+		if _, err := datastore.PutMulti(p.Context, keys, toSave); err != nil {
+			log.Errorf(p.Context, "datastore.PutMulti(..., %+v, %+v): %v; hope datastore gets fixed", keys, toSave, err)
+			return false, err
+		}
+
+		// Notify everyone.
+		allMembers := []string{}
+		for _, nat := range p.Variant.Nations {
+			allMembers = append(allMembers, string(nat))
+		}
+		notificationBody := fmt.Sprintf(
+			"%v weren't ready to resolve and gave no orders, since they still have grace periods phase resolution has been postponed %v (until %v).",
```

Comment:
The English will still be a little off here for a single nation: "England weren't ready"

Reply:
%v missed the last deadline. They are using one of their grace periods of %v. The new deadline for this phase willb e %v.

Reply:
I think that sounds better and explains it a bit clearer.

Reply:
Problem is that multiple users will be using their grace periods. How would you phrase that?

Reply:
@tttppp I changed the pluralization to be correct.